src/urllib3/contrib/emscripten/fetch.py

```diff
@@ -403,6 +534,145 @@ def send_request(request: EmscriptenRequest) -> EmscriptenResponse:
             raise _RequestError(err.message, request=request)
 
 
+def send_jspi_request(
+    request: EmscriptenRequest, streaming: bool
+) -> EmscriptenResponse:
+    """Send a request using Webassembly Javascript Promise Integration (experimental)
+       to wrap the asynchronous javascript fetch api.
+
+    Args:
+        request (EmscriptenRequest): Request to send
+        streaming : Whether to stream response
+    """
+    timeout = request.timeout
+    js_abort_controller = js.AbortController.new()
+    headers = {k: v for k, v in request.headers.items() if k not in HEADERS_TO_IGNORE}
+    req_body = request.body
+    fetch_data = {
+        "headers": headers,
+        "body": to_js(req_body),
+        "method": request.method,
+        "signal": js_abort_controller.signal,
+    }
+    # Call javascript fetch (async api, returns a promise)
+    fetcher_promise_js = js.fetch(request.url, _obj_from_dict(fetch_data))
+    # Now suspend webassembly until we resolve that promise
+    # or time out.
+    response_js = _run_sync_with_timeout(
+        fetcher_promise_js,
+        timeout,
+        js_abort_controller,
+        request=request,
+        response=None,
+    )
+    headers = {}
+    header_iter = response_js.headers.entries()
+    while True:
+        iter_value_js = header_iter.next()
+        if getattr(iter_value_js, "done", False):
+            break
+        else:
+            headers[str(iter_value_js.value[0])] = str(iter_value_js.value[1])
+    status_code = response_js.status
+    body: bytes | io.RawIOBase = b""
+
+    response = EmscriptenResponse(
+        status_code=status_code, headers=headers, body=b"", request=request
+    )
+    if streaming:
+        # get via inputstream
+        if response_js.body is not None:
+            # get a reader from the fetch response
+            body_stream_js = response_js.body.getReader()
+            body = _JSPIReadStream(
+                body_stream_js, timeout, request, response, js_abort_controller
+            )
+    else:
+        # get directly via arraybuffer
+        # n.b. this is another async Javascript call.
+        body = _run_sync_with_timeout(
+            response_js.arrayBuffer(),
+            timeout,
+            js_abort_controller,
+            request=request,
+            response=response,
+        ).to_py()
```

Comment:
The `arrayBuffer()` call isn't cancelable so I think we want to just use `run_sync` directly here.
```suggestion
        body = run_sync(response_js.arrayBuffer()).to_py()
```

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/Response/arrayBuffer says it can return AbortError, so I figure at least mozilla think that it can be cancelled between header and response fetch, although I haven't delved into browser code.

Reply:
Cool. Indeed if you poke around the fetch spec long enough you can find that the spec says this should happen:
> To abort a fetch() call with a promise, request, responseObject, and an error:
> ...
> 5. If response’s [body](https://fetch.spec.whatwg.org/#concept-response-body) is non-null and is [readable](https://streams.spec.whatwg.org/#readablestream-readable), then [error](https://streams.spec.whatwg.org/#readablestream-error) response’s [body](https://fetch.spec.whatwg.org/#concept-response-body) with error.


https://fetch.spec.whatwg.org/#abort-fetch